File: public/offline-template/start-server.py
#!/usr/bin/env python3
"""
Simple HTTP server for Mermaid Slides offline package.
Works with Python 3.x
"""

import http.server
import socketserver
import webbrowser
import os
import sys
import argparse
from pathlib import Path

# Configuration
PORT = 3000
HOST = '0.0.0.0'  # Bind to all interfaces for Docker compatibility

def find_available_port(start_port=3000):
    """Find an available port starting from the given port."""
    import socket
    port = start_port
    while port < start_port + 100:  # Try up to 100 ports
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                s.bind((HOST, port))
                return port
        except OSError:
            port += 1
    return None

def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser(description='Mermaid Slides Offline Server')
    parser.add_argument('-p', '--port', type=int, default=PORT, 
                       help=f'Port to run the server on (default: {PORT})')
    parser.add_argument('--no-browser', action='store_true',
                       help='Do not automatically open browser')
    args = parser.parse_args()

    # Change to the directory containing this script
    script_dir = Path(__file__).parent
    os.chdir(script_dir)

    print("🧜‍♀️ Mermaid Slides - Offline Server")
    print("=" * 40)

    # Find available port
    port = find_available_port(args.port)
    if port is None:
        print(f"❌ Error: Could not find an available port starting from {args.port}")
        sys.exit(1)

    # Set up the server
    handler = http.server.SimpleHTTPRequestHandler

    try:
        with socketserver.TCPServer((HOST, port), handler) as httpd:
            url = f"http://{HOST}:{port}"
            print(f"✅ Server running at: {url}")
            print(f"📁 Serving files from: {script_dir}")
            print("\n🚀 Opening in your default browser...")
            print("💡 Press Ctrl+C to stop the server\n")

            # Open browser (unless disabled)
            if not args.no_browser:
                webbrowser.open(url)

            # Start serving
            httpd.serve_forever()

    except KeyboardInterrupt:
        print("\n👋 Server stopped. Thanks for using Mermaid Slides!")
    except Exception as e:
        print(f"❌ Error starting server: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: public/offline-template/start-server.js
#!/usr/bin/env node
const http = require('http');
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');

// Parse command line arguments
const args = process.argv.slice(2);
let PORT = 3000;
let noBrowser = false;

for (let i = 0; i < args.length; i++) {
    if (args[i] === '-p' || args[i] === '--port') {
        PORT = parseInt(args[i + 1]);
        i++; // Skip next argument
    } else if (args[i] === '--no-browser') {
        noBrowser = true;
    } else if (args[i] === '-h' || args[i] === '--help') {
        console.log('Mermaid Slides Offline Server');
        console.log('Usage: node start-server.js [options]');
        console.log('Options:');
        console.log('  -p, --port <number>  Port to run the server on (default: 3000)');
        console.log('  --no-browser         Do not automatically open browser');
        console.log('  -h, --help           Show this help message');
        process.exit(0);
    }
}

const HOST = '0.0.0.0';

const mimeTypes = {
    '.html': 'text/html',
    '.js': 'text/javascript',
    '.css': 'text/css',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpg',
    '.gif': 'image/gif',
    '.svg': 'image/svg+xml',
    '.wav': 'audio/wav',
    '.mp4': 'video/mp4',
    '.woff': 'application/font-woff',
    '.ttf': 'application/font-ttf',
    '.eot': 'application/vnd.ms-fontobject',
    '.otf': 'application/font-otf',
    '.wasm': 'application/wasm'
};

// Everything served must live inside this directory. Requests are resolved
// against it and then checked, rather than trusted, because a request path is
// attacker-controlled input: `/../../../etc/passwd` is a perfectly valid thing
// for a client to send.
const ROOT = path.resolve(__dirname);

/**
 * Turn a request URL into an absolute path inside ROOT, or null if it escapes.
 *
 * Returning null rather than throwing keeps the caller's error handling in one
 * place: every rejection is answered identically, so a probe cannot tell a
 * malformed path from a blocked one from a missing file.
 */
function resolveWithinRoot(requestUrl) {
    let pathname;

    try {
        // The base is a throwaway; only the pathname is used. Parsing this way
        // discards the query string and fragment, which would otherwise end up
        // in the filename and confuse the extension lookup below.
        pathname = decodeURIComponent(new URL(requestUrl, 'http://localhost').pathname);
    } catch {
        // Malformed percent-encoding, e.g. a stray '%'.
        return null;
    }

    // A null byte can truncate the path inside some system calls.
    if (pathname.indexOf('\0') !== -1) {
        return null;
    }

    // Leading '.' makes the path relative, so resolve() cannot be pushed to the
    // filesystem root by a leading slash.
    const resolved = path.resolve(ROOT, '.' + pathname);

    // The separator check matters: without it, a sibling directory whose name
    // merely starts with ROOT would pass.
    if (resolved !== ROOT && !resolved.startsWith(ROOT + path.sep)) {
        return null;
    }

    return resolved;
}

const server = http.createServer((req, res) => {
    const resolved = resolveWithinRoot(req.url);

    if (resolved === null) {
        res.writeHead(403);
        res.end('Forbidden');
        return;
    }

    fs.stat(resolved, (statErr, stats) => {
        if (statErr) {
            res.writeHead(404);
            res.end('File not found');
            return;
        }

        // A directory serves its index.html, which is also how '/' reaches the
        // application itself.
        const filePath = stats.isDirectory() ? path.join(resolved, 'index.html') : resolved;
        const extname = path.extname(filePath).toLowerCase();
        const contentType = mimeTypes[extname] || 'application/octet-stream';

        fs.readFile(filePath, (err, content) => {
            if (err) {
                res.writeHead(err.code === 'ENOENT' ? 404 : 500);
                res.end(err.code === 'ENOENT' ? 'File not found' : 'Server error');
                return;
            }

            res.writeHead(200, { 'Content-Type': contentType });
            res.end(content);
        });
    });
});

server.listen(PORT, HOST, () => {
    const url = `http://localhost:${PORT}`;
    console.log('🧜‍♀️ Mermaid Slides - Offline Server');
    console.log('='.repeat(40));
    console.log(`✅ Server running at: ${url}`);
    console.log(`📁 Serving files from: ${__dirname}`);
    console.log('\n🚀 Opening in your default browser...');
    console.log('💡 Press Ctrl+C to stop the server\n');

    // Open browser (unless disabled)
    if (!noBrowser) {
        const start = process.platform === 'darwin' ? 'open' :
                      process.platform === 'win32' ? 'start' : 'xdg-open';
        exec(`${start} ${url}`);
    }
});